### PR TITLE
AsyncSlf4jSpanObserverTest logs Zipkin-compatible spans

### PR DESCRIPTION
**Pull request by uschi2000**

### CODE REVIEW

File: tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java [hunk 1]
```diff
@@ -28,6 +35,44 @@
...
+                    .traceId(span.getTraceId())
+                    .id(span.getSpanId())
+                    .name(span.getOperation())
+                    .parentId(span.getParentSpanId())
+                    .timestamp(span.getStartTimeMs())
+                    .duration(nanoToMilli(span.getDurationNs()))  // Zipkin-durations are micro-seconds, rounded up
+                    .build();
+        }
+
+        static long nanoToMilli(long nano) {
```
Comment (markelliot) [inline tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java]: This converts to microseconds, right?

**Review by markelliot — CHANGES_REQUESTED** [review]

File: tracing/src/main/java/com/palantir/remoting1/tracing/OpenSpan.java [hunk 2]
```diff
@@ -33,20 +33,20 @@
...
 
     /**
-     * Returns the start time in milliseconds of the span represented by this state.
+     * Returns the start time in microseconds since epoch start of the span represented by this state.
      * <p>
      * Users of this class should not set this value manually in the builder, it is configured
      * automatically when using the {@link #builder()} static.
      */
-    public abstract long getStartTimeMs();
+    public abstract long getStartTimeMicroSeconds();
```
Comment (markelliot) [inline tracing/src/main/java/com/palantir/remoting1/tracing/OpenSpan.java:41]: probably worth modifying Span to match

File: tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java [hunk 3]
```diff
@@ -56,11 +56,11 @@ static ZipkinCompatibleSerializableSpan fromSpan(Span span) {
...
                     .parentId(span.getParentSpanId())
                     .timestamp(span.getStartTimeMs())
-                    .duration(nanoToMilli(span.getDurationNs()))  // Zipkin-durations are micro-seconds, rounded up
+                    .duration(nanoToMicro(span.getDurationNs()))  // Zipkin-durations are micro-seconds, rounded up
                     .build();
         }
 
-        static long nanoToMilli(long nano) {
+        static long nanoToMicro(long nano) {
             return (long) Math.ceil(nano / (double) 1000);
```
Comment (schlosna) [inline tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java]: Why are we doing double math here? This should be:

`return Math.max(1L, nano / 1000L);`

File: tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java [hunk 4]
```diff
@@ -28,6 +35,44 @@
...
+                    .id(span.getSpanId())
+                    .name(span.getOperation())
+                    .parentId(span.getParentSpanId())
+                    .timestamp(span.getStartTimeMicroSeconds())
+                    .duration(nanoToMicro(span.getDurationNanoSeconds()))  // Zipkin-durations are micro-seconds, rounded up
+                    .build();
+        }
+
+        static long nanoToMicro(long nano) {
+            return (long) Math.ceil(nano / (double) 1000);
```
Comment (schlosna) [inline tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java]: Why are we doing double math here? This should be:

`return Math.max(1L, nano / 1000L);`

**Review by markelliot — APPROVED** [review]

File: tracing/src/main/java/com/palantir/remoting1/tracing/Tracers.java [hunk 5]
```diff
@@ -27,7 +28,11 @@ private Tracers() {}
 
     /** Returns a random ID suitable for span and trace IDs. */
     public static String randomId() {
-        return Long.toHexString(ThreadLocalRandom.current().nextLong());
+        return longToPaddedHex(ThreadLocalRandom.current().nextLong());
+    }
+
+    static String longToPaddedHex(long number) {
+        return Strings.padStart(Long.toHexString(number), 16, '0');
```
Comment (schlosna) [inline tracing/src/main/java/com/palantir/remoting1/tracing/Tracers.java:35]: Curious why we're padding

File: tracing/src/test/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserverTest.java [hunk 6]
```diff
@@ -65,17 +67,51 @@ public void after() {
...
+                .duration(44L)  // micro-seconds, rounded up
+                .build();
+        String expectedString = new String(Codec.JSON.writeSpan(zipkinSpan), StandardCharsets.UTF_8);
+        String actualString = AsyncSlf4jSpanObserver.ZipkinCompatibleSerializableSpan.fromSpan(span).toJson();
+        assertThat(actualString).isEqualTo(expectedString);
+    }
+
+    @Test
+    public void testNanoToMicro() throws Exception {
+        assertThat(AsyncSlf4jSpanObserver.ZipkinCompatibleSerializableSpan.nanoToMicro(0)).isEqualTo(0);
```
Comment (schlosna) [inline tracing/src/test/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserverTest.java]: Per https://github.com/openzipkin/zipkin-api/blob/master/zipkin-api.yaml#L281 I think you have to round up to 1 microsecond

See also https://github.com/openzipkin/zipkin/issues/1174

File: tracing/src/main/java/com/palantir/remoting1/tracing/Tracers.java (hunk 5 above)
Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/Tracers.java:35]: that's what zipkin does

File: tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java (hunk 4 above)
Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java]: assume i should round up instead of truncate?

Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/AsyncSlf4jSpanObserver.java]: will add the max

File: tracing/src/main/java/com/palantir/remoting1/tracing/OpenSpan.java (hunk 2 above)
Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/OpenSpan.java:41]: done

**Comment by uschi2000** [comment]
done

**Comment by uschi2000** [comment]
correct.

**Comment by uschi2000** [comment]
Pushed.

**Comment by uschi2000** [comment]
And yet, it seems that we're making milli seconds in places. will have to
convert more. Do not merge.

**Comment by uschi2000** [comment]
(And double-checked that zipkin wants micro, not milli)

**Pull request merged**